In [2]:
import pandas as pd
import numpy as np

## NYC  crashes

In [3]:
nyc_df = pd.read_csv('nyc_crashes.csv', header=0)

C:\Users\tymek\AppData\Local\Temp\ipykernel_26172\261270579.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  nyc_df = pd.read_csv('nyc_crashes.csv', header=0)


In [4]:
nyc_df['ZIP CODE'] = nyc_df['ZIP CODE'].fillna(0).astype(str).str.strip().apply(lambda x: '0' if x == '' else x).astype(float).astype(int).astype(str)

In [152]:
nyc_df.tail()

CRASH DATE CRASH TIME    BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
2075956  03/24/2024      15:02   BROOKLYN    11207  40.660410 -73.896320   
2075957  03/13/2024      10:00  MANHATTAN    10032  40.833660 -73.940040   
2075958  03/15/2024      19:00      BRONX    10467  40.881073 -73.878494   
2075959  03/22/2024      17:30   BROOKLYN    11226  40.654880 -73.961890   
2075960  03/24/2024      13:05      BRONX    10455  40.817368 -73.902590   

                        LOCATION     ON STREET NAME   CROSS STREET NAME  \
2075956    (40.66041, -73.89632)    NEW LOTS AVENUE        MALTA STREET   
2075957    (40.83366, -73.94004)                NaN                 NaN   
2075958  (40.881073, -73.878494)  BAINBRIDGE AVENUE  EAST GUN HILL ROAD   
2075959    (40.65488, -73.96189)       OCEAN AVENUE     PARKSIDE AVENUE   
2075960   (40.817368, -73.90259)    EAST 156 STREET     PROSPECT AVENUE   

                   OFF STREET NAME  ...  CONTRIBUTING FACTOR VEHICLE 2  \
2075956                        NaN  ...    Traffic Control Disregarded   
2075957  461       WEST 159 STREET  ...                    Unspecified   
2075958                        NaN  ...                    Unspecified   
2075959                        NaN  ...                    Unspecified   
2075960                        NaN  ...                    Unspecified   

         CONTRIBUTING FACTOR VEHICLE 3  CONTRIBUTING FACTOR VEHICLE 4  \
2075956                    Unspecified                    Unspecified   
2075957                            NaN                            NaN   
2075958                            NaN                            NaN   
2075959                            NaN                            NaN   
2075960                            NaN                            NaN   

         CONTRIBUTING FACTOR VEHICLE 5  COLLISION_ID  \
2075956                            NaN       4712169   
2075957                            NaN       4712585   
2075958                            NaN       4712596   
2075959                            NaN       4712610   
2075960                            NaN       4712207   

                         VEHICLE TYPE CODE 1  \
2075956                                Sedan   
2075957                                Sedan   
2075958                        Pick-up Truck   
2075959  Station Wagon/Sport Utility Vehicle   
2075960  Station Wagon/Sport Utility Vehicle   

                         VEHICLE TYPE CODE 2  VEHICLE TYPE CODE 3  \
2075956                        Pick-up Truck                Sedan   
2075957                                Sedan                  NaN   
2075958  Station Wagon/Sport Utility Vehicle                  NaN   
2075959                                 Dump                  NaN   
2075960                                  NaN                  NaN   

        VEHICLE TYPE CODE 4 VEHICLE TYPE CODE 5  
2075956               Sedan                 NaN  
2075957                 NaN                 NaN  
2075958                 NaN                 NaN  
2075959                 NaN                 NaN  
2075960                 NaN                 NaN  

[5 rows x 29 columns]

## NYC vehicles

In [5]:
nyc_vehicles = pd.read_csv('nyc_vehicles.csv', header=0)

C:\Users\tymek\AppData\Local\Temp\ipykernel_26172\3742854358.py:1: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  nyc_vehicles = pd.read_csv('nyc_vehicles.csv', header=0)


In [143]:
nyc_vehicles.tail(5)

UNIQUE_ID  COLLISION_ID  CRASH_DATE CRASH_TIME  \
4166566   20622266       4712130  03/24/2024      17:40   
4166567   20622808       4712209  03/24/2024      16:32   
4166568   20622933       4712118  03/24/2024      20:19   
4166569   20622816       4712072  03/24/2024       3:03   
4166570   20623712       4705579  02/27/2024       7:52   

                                   VEHICLE_ID STATE_REGISTRATION VEHICLE_TYPE  \
4166566  593e8074-ba36-4f47-91c0-e5e1fcb85d70                 NY    Ambulance   
4166567  806c38a9-0bf5-4f2e-ad23-fb2c81854d4c                 NY          NaN   
4166568  e74a622a-1465-4f68-8f65-b600c216ebdf                 NY        Sedan   
4166569  658a3f4a-5d72-41ed-938a-03010b0401ff                 NY    Box Truck   
4166570  5bafb1e9-26b9-493b-be04-5bfe7d5c3bd5                NaN         Bike   

           VEHICLE_MAKE VEHICLE_MODEL  VEHICLE_YEAR  ...  \
4166566  FORD-TRUCK/BUS           NaN        2022.0  ...   
4166567   DODG -CAR/SUV           NaN        2021.0  ...   
4166568   HOND -CAR/SUV           NaN        2019.0  ...   
4166569  FRHT-TRUCK/BUS           NaN        2004.0  ...   
4166570             NaN           NaN           NaN  ...   

                    PRE_CRASH    POINT_OF_IMPACT     VEHICLE_DAMAGE  \
4166566         Making U Turn  Right Rear Bumper  Right Rear Bumper   
4166567                   NaN                NaN                NaN   
4166568      Making Left Turn    Left Side Doors    Left Side Doors   
4166569                Parked    Center Back End    Center Back End   
4166570  Going Straight Ahead              Other              Other   

        VEHICLE_DAMAGE_1 VEHICLE_DAMAGE_2 VEHICLE_DAMAGE_3  \
4166566        No Damage        No Damage        No Damage   
4166567              NaN              NaN              NaN   
4166568              NaN              NaN              NaN   
4166569              NaN              NaN              NaN   
4166570              NaN              NaN              NaN   

        PUBLIC_PROPERTY_DAMAGE PUBLIC_PROPERTY_DAMAGE_TYPE  \
4166566                      N                         NaN   
4166567            Unspecified                         NaN   
4166568                      N                         NaN   
4166569                      N                         NaN   
4166570                      N                         NaN   

        CONTRIBUTING_FACTOR_1 CONTRIBUTING_FACTOR_2  
4166566           Unspecified           Unspecified  
4166567                   NaN                   NaN  
4166568           Unspecified           Unspecified  
4166569           Unspecified           Unspecified  
4166570           Unspecified           Unspecified  

[5 rows x 25 columns]

## Vehicle specs

In [144]:
vehicles = pd.read_csv('vehicles.csv', header=0)

C:\Users\tymek\AppData\Local\Temp\ipykernel_18924\4012450446.py:1: DtypeWarning: Columns (72,74,75,77) have mixed types. Specify dtype option on import or set low_memory=False.
  vehicles = pd.read_csv('vehicles.csv', header=0)


In [145]:
vehicles.columns

Index(['barrels08', 'barrelsA08', 'charge120', 'charge240', 'city08',
       'city08U', 'cityA08', 'cityA08U', 'cityCD', 'cityE', 'cityUF', 'co2',
       'co2A', 'co2TailpipeAGpm', 'co2TailpipeGpm', 'comb08', 'comb08U',
       'combA08', 'combA08U', 'combE', 'combinedCD', 'combinedUF', 'cylinders',
       'displ', 'drive', 'engId', 'eng_dscr', 'feScore', 'fuelCost08',
       'fuelCostA08', 'fuelType', 'fuelType1', 'ghgScore', 'ghgScoreA',
       'highway08', 'highway08U', 'highwayA08', 'highwayA08U', 'highwayCD',
       'highwayE', 'highwayUF', 'hlv', 'hpv', 'id', 'lv2', 'lv4', 'make',
       'model', 'mpgData', 'phevBlended', 'pv2', 'pv4', 'range', 'rangeCity',
       'rangeCityA', 'rangeHwy', 'rangeHwyA', 'trany', 'UCity', 'UCityA',
       'UHighway', 'UHighwayA', 'VClass', 'year', 'youSaveSpend', 'baseModel',
       'guzzler', 'trans_dscr', 'tCharger', 'sCharger', 'atvType', 'fuelType2',
       'rangeA', 'evMotor', 'mfrCode', 'c240Dscr', 'charge240b', 'c240bDscr',
       'createdOn'

In [147]:
vehicles_small = vehicles[['make', 'model', 'year', 'VClass', 'displ', 'cylinders', 'drive', 'eng_dscr']]

In [148]:
vehicles_small[(vehicles_small['make']=='Hyundai') & (vehicles_small['year']==2022) & (vehicles_small['model'].str.contains(' N'))] 

make       model  year                           VClass  displ  \
37184  Hyundai  Veloster N  2022                     Compact Cars    2.0   
37185  Hyundai  Veloster N  2022                     Compact Cars    2.0   
37873  Hyundai   Elantra N  2022                     Midsize Cars    2.0   
37874  Hyundai   Elantra N  2022                     Midsize Cars    2.0   
38001  Hyundai      Kona N  2022  Small Sport Utility Vehicle 2WD    2.0   

       cylinders              drive eng_dscr  
37184        4.0  Front-Wheel Drive     SIDI  
37185        4.0  Front-Wheel Drive     SIDI  
37873        4.0  Front-Wheel Drive     SIDI  
37874        4.0  Front-Wheel Drive     SIDI  
38001        4.0  Front-Wheel Drive     SIDI

## Community data

In [6]:
race_df = pd.read_csv('uhf42_race.csv', header=0)

In [150]:
race_df.head()

TimePeriod GeoTypeDesc  GeoID  GeoRank                Geography  \
0    2017-21      UHF 42    101        4  Kingsbridge - Riverdale   
1    2017-21      UHF 42    102        4          Northeast Bronx   
2    2017-21      UHF 42    103        4       Fordham - Bronx Pk   
3    2017-21      UHF 42    104        4     Pelham - Throgs Neck   
4    2017-21      UHF 42    105        4         Crotona -Tremont   

   Asian alone (percent)  Black alone (percent)  Hispanic alone (percent)  \
0                    3.7                   11.3                      45.6   
1                    3.5                   56.6                      27.7   
2                    4.2                   20.2                      66.8   
3                    8.4                   19.4                      53.3   
4                    1.3                   27.1                      67.0   

   Non-white (percent) Total population  White alone (percent)  
0                 64.2           96,643                   35.8  
1                 91.8          212,012                    8.2  
2                 93.5          271,446                    6.5  
3                 84.4          308,864                   15.6  
4                 97.7          219,815                    2.3

## Zipcodes

In [7]:
import pandas as pd

data = {
    'Zip Code': ['10463, 10471', '10466, 10469, 10470, 10475', '10458, 10467, 10468', '10461, 10462, 10464, 10465, 10472, 10473', '10453, 10457, 10460', '10451, 10452, 10456', '10454, 10455, 10459, 10474', '11211, 11222', '11201, 11205, 11215, 11217, 11231', '11213, 11212, 11216, 11233, 11238', '11207, 11208', '11220, 11232', '11204, 11218, 11219, 11230', '11203, 11210, 11225, 11226', '11234, 11236, 11239', '11209, 11214, 11228', '11223, 11224, 11229, 11235', '11206, 11221, 11237', '10031, 10032, 10033, 10034, 10040', '10026, 10027, 10030, 10037, 10039', '10029, 10035', '10023, 10024, 10025', '10021, 10028, 10044, 10128', '10001, 10011, 10018, 10019, 10020, 10036', '10010, 10016, 10017, 10022', '10012, 10013, 10014', '10002, 10003, 10009', '10004, 10005, 10006, 10007, 10038, 10280', '11101, 11102, 11103, 11104, 11105, 11106', '11368, 11369, 11370, 11372, 11373, 11377, 11378', '11354, 11355, 11356, 11357, 11358, 11359, 11360', '11361, 11362, 11363, 11364', '11374, 11375, 11379, 11385', '11365, 11366, 11367', '11414, 11415, 11416, 11417, 11418, 11419, 11420, 11421', '11412, 11423, 11432, 11433, 11434, 11435, 11436', '11004, 11005, 11411, 11413, 11422, 11426, 11427, 11428, 11429', '11691, 11692, 11693, 11694, 11695, 11697', '10302, 10303, 10310', '10301, 10304, 10305', '10314', '10306, 10307, 10308, 10309, 10312'],
    'District Name': ['Kingsbridge - Riverdale', 'Northeast Bronx', 'Fordham - Bronx Park', 'Pelham - Throgs Neck', 'Crotona - Tremont', 'High Bridge - Morrisania', 'Hunts Point - Mott Haven', 'Greenpoint', 'Downtown - Heights - Park Slope', 'Bedford Stuyvesant - Crown Heights', 'East New York', 'Sunset Park', 'Borough Park', 'East Flatbush - Flatbush', 'Canarsie - Flatlands', 'Bensonhurst - Bay Ridge', 'Coney Island - Sheepshead Bay', 'Williamsburg - Bushwick', 'Washington Heights - Inwood', 'Central Harlem - Morningside Heights', 'East Harlem', 'Upper West Side', 'Upper East Side', 'Chelsea - Clinton', 'Gramercy Park - Murray Hill', 'Greenwich Village - SoHo', 'Union Square - Lower East Side', 'Lower Manhattan', 'Long Island City - Astoria', 'West Queens', 'Flushing - Clearview', 'Bayside - Little Neck', 'Ridgewood - Forest Hills', 'Fresh Meadows', 'Southwest Queens', 'Jamaica', 'Southeast Queens', 'Rockaway', 'Port Richmond', 'Stapleton - St. George', 'Willowbrook', 'South Beach - Tottenville'],
    'District ID': [101, 102, 103, 104, 105, 106, 107, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 501, 502, 503, 504]
}

df_zip = pd.DataFrame(data)

In [8]:
df_zip = df_zip.set_index(['District Name', 'District ID']).apply(lambda x: x.str.split(', ').explode()).reset_index()

In [9]:
df_zip

District Name  District ID Zip Code
0      Kingsbridge - Riverdale          101    10463
1      Kingsbridge - Riverdale          101    10471
2              Northeast Bronx          102    10466
3              Northeast Bronx          102    10469
4              Northeast Bronx          102    10470
..                         ...          ...      ...
171  South Beach - Tottenville          504    10306
172  South Beach - Tottenville          504    10307
173  South Beach - Tottenville          504    10308
174  South Beach - Tottenville          504    10309
175  South Beach - Tottenville          504    10312

[176 rows x 3 columns]

In [156]:
len(np.unique(small_nyc['ZIP CODE']))

235

w danych z wypadkami jest więcej kodów pocztowych niż mapperze zipcode - district

### joining

In [212]:
nyc_df.columns

Index(['CRASH DATE', 'CRASH TIME', 'BOROUGH', 'ZIP CODE', 'LATITUDE',
       'LONGITUDE', 'LOCATION', 'ON STREET NAME', 'CROSS STREET NAME',
       'OFF STREET NAME', 'NUMBER OF PERSONS INJURED',
       'NUMBER OF PERSONS KILLED', 'NUMBER OF PEDESTRIANS INJURED',
       'NUMBER OF PEDESTRIANS KILLED', 'NUMBER OF CYCLIST INJURED',
       'NUMBER OF CYCLIST KILLED', 'NUMBER OF MOTORIST INJURED',
       'NUMBER OF MOTORIST KILLED', 'CONTRIBUTING FACTOR VEHICLE 1',
       'CONTRIBUTING FACTOR VEHICLE 2', 'CONTRIBUTING FACTOR VEHICLE 3',
       'CONTRIBUTING FACTOR VEHICLE 4', 'CONTRIBUTING FACTOR VEHICLE 5',
       'COLLISION_ID', 'VEHICLE TYPE CODE 1', 'VEHICLE TYPE CODE 2',
       'VEHICLE TYPE CODE 3', 'VEHICLE TYPE CODE 4', 'VEHICLE TYPE CODE 5'],
      dtype='object')

In [10]:
small_nyc = nyc_df[['COLLISION_ID', 'CRASH DATE', 'CRASH TIME', 'BOROUGH', 'ZIP CODE', 'VEHICLE TYPE CODE 1']] # mała ramka do testów

In [214]:
len(small_nyc)  # liczba wypadkow

2075961

In [215]:
outer = pd.merge(small_nyc, df_zip, how='outer', left_on='ZIP CODE', right_on='Zip Code', indicator=True)

In [216]:
len(outer[outer['ZIP CODE'] == '0'])

646218

te co nie mają pokrycia to w większości 0, czyli nie wiadomo gdzie zdarzył się wypadek

In [217]:
excluded_rows = outer[(outer['_merge'] == 'left_only') & (outer['ZIP CODE'] != '0')]

In [218]:
len(excluded_rows)

22389

In [219]:
np.unique(df_zip['Zip Code'])

array(['10001', '10002', '10003', '10004', '10005', '10006', '10007',
       '10009', '10010', '10011', '10012', '10013', '10014', '10016',
       '10017', '10018', '10019', '10020', '10021', '10022', '10023',
       '10024', '10025', '10026', '10027', '10028', '10029', '10030',
       '10031', '10032', '10033', '10034', '10035', '10036', '10037',
       '10038', '10039', '10040', '10044', '10128', '10280', '10301',
       '10302', '10303', '10304', '10305', '10306', '10307', '10308',
       '10309', '10310', '10312', '10314', '10451', '10452', '10453',
       '10454', '10455', '10456', '10457', '10458', '10459', '10460',
       '10461', '10462', '10463', '10464', '10465', '10466', '10467',
       '10468', '10469', '10470', '10471', '10472', '10473', '10474',
       '10475', '11004', '11005', '11101', '11102', '11103', '11104',
       '11105', '11106', '11201', '11203', '11204', '11205', '11206',
       '11207', '11208', '11209', '11210', '11211', '11212', '11213',
       '11214', '112

In [220]:
np.unique(excluded_rows['ZIP CODE'])

array(['10000', '10041', '10045', '10048', '10055', '10065', '10069',
       '10075', '10103', '10105', '10106', '10107', '10110', '10111',
       '10112', '10115', '10118', '10119', '10120', '10121', '10122',
       '10123', '10151', '10152', '10153', '10154', '10155', '10158',
       '10162', '10165', '10166', '10167', '10168', '10169', '10170',
       '10171', '10172', '10173', '10174', '10176', '10177', '10178',
       '10179', '10271', '10278', '10279', '10281', '10282', '10803',
       '11001', '11040', '11109', '11241', '11242', '11249', '11251',
       '11371', '11430'], dtype=object)

niektórych kodów pocztowych nie ma w danych. to zapewne dlatego że niektóre budynki w NYC mają swoje własne zip codes, 
więc nijak się odnoszą do obszarów: "10041; it was one of 41 buildings in Manhattan that had their own ZIP Codes as of 2019". Jest ich tylko 22389 (1% danych), więc zrobimy left joina i wrzucimy je po prostu wrzucić do 'Unknown'.

In [11]:
inner = pd.merge(small_nyc, df_zip, how='left', left_on='ZIP CODE', right_on='Zip Code')

In [222]:
len(inner)

2075961

In [223]:
inner[inner['Zip Code'].isna()]

COLLISION_ID  CRASH DATE CRASH TIME BOROUGH ZIP CODE  \
0             4455765  09/11/2021       2:39     NaN        0   
1             4513547  03/26/2022      11:45     NaN        0   
2             4541903  06/29/2022       6:55     NaN        0   
5             4407458  04/14/2021      12:47     NaN        0   
6             4486555  12/14/2021      17:05     NaN        0   
...               ...         ...        ...     ...      ...   
2075947       4712126  03/24/2024      20:40     NaN        0   
2075948       4712117  03/24/2024       6:53     NaN        0   
2075952       4712233  03/24/2024      13:15     NaN        0   
2075953       4712656  03/10/2024      11:45     NaN        0   
2075955       4712264  03/24/2024       6:46     NaN        0   

                         VEHICLE TYPE CODE 1 District Name  District ID  \
0                                      Sedan           NaN          NaN   
1                                      Sedan           NaN          NaN   
2                                      Sedan           NaN          NaN   
5                                       Dump           NaN          NaN   
6                                      Sedan           NaN          NaN   
...                                      ...           ...          ...   
2075947                        Pick-up Truck           NaN          NaN   
2075948  Station Wagon/Sport Utility Vehicle           NaN          NaN   
2075952  Station Wagon/Sport Utility Vehicle           NaN          NaN   
2075953                                Sedan           NaN          NaN   
2075955                                Sedan           NaN          NaN   

        Zip Code  
0            NaN  
1            NaN  
2            NaN  
5            NaN  
6            NaN  
...          ...  
2075947      NaN  
2075948      NaN  
2075952      NaN  
2075953      NaN  
2075955      NaN  

[668607 rows x 9 columns]

teraz można do tego dodać dane rasowe

In [12]:
crash_race = pd.merge(inner, race_df, how='left', left_on='District ID', right_on='GeoID')

In [13]:
crash_race.tail(5)

COLLISION_ID  CRASH DATE CRASH TIME    BOROUGH ZIP CODE  \
2075956       4712169  03/24/2024      15:02   BROOKLYN    11207   
2075957       4712585  03/13/2024      10:00  MANHATTAN    10032   
2075958       4712596  03/15/2024      19:00      BRONX    10467   
2075959       4712610  03/22/2024      17:30   BROOKLYN    11226   
2075960       4712207  03/24/2024      13:05      BRONX    10455   

                         VEHICLE TYPE CODE 1                District Name  \
2075956                                Sedan                East New York   
2075957                                Sedan  Washington Heights - Inwood   
2075958                        Pick-up Truck         Fordham - Bronx Park   
2075959  Station Wagon/Sport Utility Vehicle     East Flatbush - Flatbush   
2075960  Station Wagon/Sport Utility Vehicle     Hunts Point - Mott Haven   

         District ID Zip Code TimePeriod GeoTypeDesc  GeoID  GeoRank  \
2075956        204.0    11207    2017-21      UHF 42  204.0      4.0   
2075957        301.0    10032    2017-21      UHF 42  301.0      4.0   
2075958        103.0    10467    2017-21      UHF 42  103.0      4.0   
2075959        207.0    11226    2017-21      UHF 42  207.0      4.0   
2075960        107.0    10455    2017-21      UHF 42  107.0      4.0   

                        Geography  Asian alone (percent)  \
2075956             East New York                    4.4   
2075957        Washington Heights                    3.4   
2075958        Fordham - Bronx Pk                    4.2   
2075959  East Flatbush - Flatbush                    3.0   
2075960  Hunts Point - Mott Haven                    0.4   

         Black alone (percent)  Hispanic alone (percent)  Non-white (percent)  \
2075956                   50.5                      37.7                 96.1   
2075957                   10.6                      63.9                 80.7   
2075958                   20.2                      66.8                 93.5   
2075959                   64.5                      10.2                 82.1   
2075960                   26.4                      69.8                 97.7   

        Total population  White alone (percent)  
2075956          205,414                    3.9  
2075957          272,205                   19.3  
2075958          271,446                    6.5  
2075959          304,594                   17.9  
2075960          144,304                    2.3

żeby nie zmieniać rozmiaru danych możemy zmergować tylko główny pojazd biorący udział w wypadku

In [15]:
nyc_vehicles['VEHICLE_MAKE'].astype(str).value_counts()

VEHICLE_MAKE
nan                   1878022
TOYT -CAR/SUV          391897
HOND -CAR/SUV          282185
NISS -CAR/SUV          232333
FORD -CAR/SUV          192819
                       ...   
MACK / TRAC                 1
SAGE                        1
UOGUE                       1
ORIIN                       1
USPS Grumman Truck          1
Name: count, Length: 12758, dtype: int64

In [18]:
# weźmy jeden pojazd na jedną kolizję
nyc_vehicles['VEHICLE_MAKE'].fillna('', inplace=True)

def longest_name(series):
    longest = series.str.len().idxmax()
    return series.loc[longest]

result = nyc_vehicles.groupby('COLLISION_ID')['VEHICLE_MAKE'].agg(longest_name)

In [22]:
result.reset_index()['VEHICLE_MAKE'].value_counts().head(50)

VEHICLE_MAKE
                   803698
TOYT -CAR/SUV      227297
HOND -CAR/SUV      163025
NISS -CAR/SUV      134573
FORD -CAR/SUV      110686
CHEV -CAR/SUV       63403
HYUN -CAR/SUV       47217
MERZ -CAR/SUV       43767
JEEP -CAR/SUV       42516
DODG -CAR/SUV       35114
LEXS -CAR/SUV       30592
ACUR -CAR/SUV       27880
INFI -CAR/SUV       25231
FORD-TRUCK/BUS      24323
FRHT-TRUCK/BUS      21461
VOLK -CAR/SUV       19078
SUBA -CAR/SUV       18540
BMW -CAR/SUV        17081
CHRY -CAR/SUV       16857
AUDI -CAR/SUV       15220
MAZD -CAR/SUV       14863
INTL-TRUCK/BUS      13003
LINC -CAR/SUV       12704
CADI -CAR/SUV       10581
MACK-TRUCK/BUS       9830
HINO -TRUCK/BUS      9081
MITS -CAR/SUV        8916
KIA -CAR/SUV         7228
VOLV -CAR/SUV        5785
LNDR -CAR/SUV        5510
CHEV-TRUCK/BUS       5243
GMC -CAR/SUV         5204
BUIC -CAR/SUV        4745
PTRB-TRUCK/BUS       4560
ISU-TRUCK/BUS        3880
MERC -CAR/SUV        3783
VOLV -TRUCK/BUS      3753
TESL -CAR/SUV        2614

chujowy format, jeśli chcemy używać samochodów będzie potrzebne dużo ręcznej roboty (w dodatku większość to i tak nulle), plus prawie wszystkie modele to nulle :(

In [16]:
len(np.unique(nyc_vehicles['COLLISION_ID']))

2074033

In [26]:
nyc_vehicles_sm = nyc_vehicles[['COLLISION_ID', 'VEHICLE_ID', 'VEHICLE_MAKE','VEHICLE_MODEL','VEHICLE_YEAR','CONTRIBUTING_FACTOR_1']]

In [27]:
fact_table = pd.merge(crash_race, nyc_vehicles_sm, how='left', on='COLLISION_ID')

In [28]:
fact_table.tail()

COLLISION_ID  CRASH DATE CRASH TIME   BOROUGH ZIP CODE  \
4168494       4712596  03/15/2024      19:00     BRONX    10467   
4168495       4712610  03/22/2024      17:30  BROOKLYN    11226   
4168496       4712610  03/22/2024      17:30  BROOKLYN    11226   
4168497       4712207  03/24/2024      13:05     BRONX    10455   
4168498       4712207  03/24/2024      13:05     BRONX    10455   

                         VEHICLE TYPE CODE 1             District Name  \
4168494                        Pick-up Truck      Fordham - Bronx Park   
4168495  Station Wagon/Sport Utility Vehicle  East Flatbush - Flatbush   
4168496  Station Wagon/Sport Utility Vehicle  East Flatbush - Flatbush   
4168497  Station Wagon/Sport Utility Vehicle  Hunts Point - Mott Haven   
4168498  Station Wagon/Sport Utility Vehicle  Hunts Point - Mott Haven   

         District ID Zip Code TimePeriod  ... Black alone (percent)  \
4168494        103.0    10467    2017-21  ...                  20.2   
4168495        207.0    11226    2017-21  ...                  64.5   
4168496        207.0    11226    2017-21  ...                  64.5   
4168497        107.0    10455    2017-21  ...                  26.4   
4168498        107.0    10455    2017-21  ...                  26.4   

         Hispanic alone (percent)  Non-white (percent) Total population  \
4168494                      66.8                 93.5          271,446   
4168495                      10.2                 82.1          304,594   
4168496                      10.2                 82.1          304,594   
4168497                      69.8                 97.7          144,304   
4168498                      69.8                 97.7          144,304   

         White alone (percent)                            VEHICLE_ID  \
4168494                    6.5  9f190717-4d0c-41b7-a6f3-4759cf34d49c   
4168495                   17.9  32e0dce3-1266-4c95-9371-b95cc89c8b62   
4168496                   17.9  99bbfe22-7cb7-47cb-976b-21f55d06bfb9   
4168497                    2.3  aaa03fb2-e0cd-402c-a127-d88229fc05fb   
4168498                    2.3  2ad3892c-4704-428e-a2d2-63590310e650   

           VEHICLE_MAKE  VEHICLE_MODEL VEHICLE_YEAR  \
4168494   DODG -CAR/SUV            NaN       2015.0   
4168495   MERZ -CAR/SUV            NaN       2014.0   
4168496  MACK-TRUCK/BUS            NaN       2022.0   
4168497   FORD -CAR/SUV            NaN       2005.0   
4168498   CHEV -CAR/SUV            NaN          NaN   

                  CONTRIBUTING_FACTOR_1  
4168494  Reaction to Uninvolved Vehicle  
4168495                     Unspecified  
4168496                     Unspecified  
4168497                     Unspecified  
4168498  Driver Inattention/Distraction  

[5 rows x 25 columns]

In [29]:
# teraz fakt to pojazd biorący udział w wypadku
len(fact_table)

4168499

In [14]:
nyc_vehicles[~nyc_vehicles['VEHICLE_MAKE'].isna()].head()

UNIQUE_ID  COLLISION_ID  CRASH_DATE CRASH_TIME  \
1    19140702       4213082  09/23/2019       8:15   
5    17044639       3434155  05/02/2016      17:35   
6    19138701       4229067  10/24/2019      13:15   
7    17303317       3503027  08/18/2016      12:39   
10   17285715       3487936  07/22/2016      15:40   

                              VEHICLE_ID STATE_REGISTRATION  \
1   0553ab4d-9500-4cba-8d98-f4d7f89d5856                 NY   
5                                 219456                 NY   
6   c53b43d9-419a-4ab1-9361-3f2979078d89                 NY   
7                                 672828                 NY   
10                                554272                 NY   

                           VEHICLE_TYPE    VEHICLE_MAKE VEHICLE_MODEL  \
1   Station Wagon/Sport Utility Vehicle   TOYT -CAR/SUV           NaN   
5                            4 dr sedan   MERZ -CAR/SUV           NaN   
6                                   Bus  FRHT-TRUCK/BUS           NaN   
7   Station Wagon/Sport Utility Vehicle   FORD -CAR/SUV           NaN   
10                          Convertible   VOLK -CAR/SUV           NaN   

    VEHICLE_YEAR  ...             PRE_CRASH           POINT_OF_IMPACT  \
1         2002.0  ...  Going Straight Ahead         Left Front Bumper   
5         2015.0  ...               Merging        Right Front Bumper   
6         2006.0  ...                Parked  Left Front Quarter Panel   
7         2005.0  ...  Going Straight Ahead          Center Front End   
10        2013.0  ...    Stopped in Traffic         Right Rear Bumper   

              VEHICLE_DAMAGE           VEHICLE_DAMAGE_1  VEHICLE_DAMAGE_2  \
1   Left Front Quarter Panel                        NaN               NaN   
5         Right Front Bumper  Right Front Quarter Panel               NaN   
6   Left Front Quarter Panel                        NaN               NaN   
7           Center Front End                  No Damage         No Damage   
10         Right Rear Bumper            Center Back End  Left Rear Bumper   

   VEHICLE_DAMAGE_3 PUBLIC_PROPERTY_DAMAGE PUBLIC_PROPERTY_DAMAGE_TYPE  \
1               NaN                      N                         NaN   
5               NaN                      N                         NaN   
6               NaN                      N                         NaN   
7         No Damage                      N                         NaN   
10              NaN                      N                         NaN   

             CONTRIBUTING_FACTOR_1 CONTRIBUTING_FACTOR_2  
1   Driver Inattention/Distraction           Unspecified  
5   Driver Inattention/Distraction  Unsafe Lane Changing  
6                      Unspecified           Unspecified  
7   Driver Inattention/Distraction           Unspecified  
10                     Unspecified           Unspecified  

[5 rows x 25 columns]

In [49]:
len(nyc_vehicles)

4166571

In [50]:
len(np.unique(nyc_vehicles['VEHICLE_ID'])) # tyle różnych samochodów brało udział w wypadkach 

2637746